### PR TITLE
[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #95657
* #95486

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 1]
```diff
@@ -185,6 +186,7 @@ inline bool check_for_seq_len_1_nested_tensor(sdp_params params, bool debug) {
 
 inline bool check_for_nested_inputs(sdp_params params, bool debug){
   if (params.query.is_nested() || params.key.is_nested() || params.value.is_nested()) {
+    // FIXME: this debug message should be removed now
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:189]: I will move this into the other PR

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 2]
```diff
@@ -56,7 +57,7 @@ inline std::array<SDPBackend, num_backends> priority_order(sdp_params params) {
   // FlashAttention parallelizes across "batch_size * num_heads"
   // MemEff parallelizes across "batch_size * num_heads * num_queries" and can
   // be more efficient. batch_size, q_len, num_heads, k = inp.query.shape
-  if (params.query.is_nested()) {
+  if (params.key.is_nested()) {
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:60]: maybe we should q,k,v looking for nested

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 3]
```diff
@@ -246,21 +248,77 @@ inline bool check_tensor_shapes(sdp_params params, bool debug) {
...
+                                                         debug);
+    }
+    return broadcastable_batch_size && broadcastable_num_heads;
+  }
 
   if (!(same_batch_size && same_num_heads)) {
     if (debug) {
       TORCH_WARN(
-        "Both fused kernels requires query, key and value to have the same batch_size and num_heads. Query.sizes(): ",
+        "For dense inputs, both fused kernels require query, key and value to have the same batch_size and num_heads. ",
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:320]: Might be worth saying, try unsqueeze and expand_to but I think that if they are savy enough to turn off the math fallback they will probably know that.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 4]
```diff
@@ -319,78 +319,234 @@ bool is_safe_to_get_storage_as_tensor(const NestedTensorImpl* tensor) {
...
+  }
 
-  Tensor cumulative_sequence_length_q = std::get<0>(cumulative_and_max_q);
-  Tensor cumulative_sequence_length_k = std::get<0>(cumulative_and_max_k);
+  int64_t max_seqlen_batch_kv = 0, Nnz_kv = 0;
+  Tensor cumulative_sequence_length_kv;
+  if (k_batch_size_needs_broadcast) {
+    if (v_batch_size_needs_broadcast) {
+      // Should not be the case that both k and v need broadcasting of batch_size
+      // i.e. output_batch_size == 1 and k/v_batch_size are both 1
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:382]: != 1 

Also is this checked in sdp_utils?

Comment (mikaylagawarecki) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:382]: Will add a check, thanks for catching

File: test/test_transformers.py [hunk 5]
```diff
@@ -1963,6 +1952,36 @@ def test_fused_kernels_seq_len_0_inputs(self):
...
+        seq_lens = torch.randint(low=1, high=32, size=(batch,)).tolist()
+        q_shape = (1, 1, num_heads, head_dim)
+        k_shape = (batch, seq_lens, num_heads, head_dim)
+        v_shape = (batch, seq_lens, 1, head_dim_v)
+
+        query = torch.randn(q_shape, device="cuda", dtype=torch.float32)
+        key = rand_nested_tensor(k_shape)
+        value = rand_nested_tensor(v_shape)
+
+        query_expanded = torch.nested.nested_tensor([query.squeeze(0) for _ in range(batch)]).transpose(1, 2)
```
Comment (drisspg) [inline test/test_transformers.py:1968]: might be nice to add a  comment of the nested shapes
ie.e # query_expended = (num_heads, 1, head_dim)

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp (hunk 4 above)
Comment (mikaylagawarecki) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:382]: edit: actually @drisspg, do you think this case output_batch_size == 1 and k/vbatch_size == 1 might potentially sense? rethinking it now I I might have missed it here cuz then it would just be

k, v `[1, num_head, x, embed_qk], [1, num_head, x, embed_v] --> [B, num_head, x, embed_qk], [B, num_head, x, embed_v]`

The case I actually missed was if 1 of k/v batch_size needs broadcast and not the other then the other must have a consistent seq_len dim, e.g. 

k, v `[1, num_head, x, embed_qk], [B, num_head, x, embed_v] --> [B, num_head, x, embed_qk], [B, num_head, x, embed_v]`

does that sound right to you?

Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:382]: To be honest this is the most logic dense part of the code and little hard to follow all the paths. But the broadcasting examples above I agree with. Like you said in the second example when k and v are nested we need to be more careful on the seq_len broadcasting

File: aten/src/ATen/native/nested/NestedTensorMatmul.cpp [hunk 6]
```diff
@@ -306,7 +306,7 @@ Tensor matmul_nested(const Tensor& self, const Tensor& mat2) {
       self_dim == 4 && self.is_contiguous() &&
       mat2_dim == 4 && mat2.is_contiguous() &&
       !(GradMode::is_enabled() && (self.requires_grad() || mat2.requires_grad()))) {
-    auto n_heads = self_sizes.select(0, 1).select(0, 0).item<int64_t>();
+    auto n_heads = self_sizes.select(0, 0).select(0, 0).item<int64_t>();
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMatmul.cpp:309]: Is this fixing a bug? If so let's split it out into a separate PR and update the corresponding matmul tests to cover it.

Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorMatmul.cpp:309]: Also can't we use opt_dim to do this?

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 7]
```diff
@@ -246,27 +314,77 @@ inline bool check_tensor_shapes(sdp_params params, bool debug) {
...
-  bool same_num_heads = params.query.is_nested()
-      ? true
-      : params.query.size(1) == params.key.size(1) &&
-          params.query.size(1) == params.value.size(1);
+  auto q_batch_size = params.query.size(0);
+  auto k_batch_size = params.key.size(0);
+  auto v_batch_size = params.value.size(0);
+
+  // Flip because check_for_nested_inputs returns false if there exists a nested input
+  bool has_nested_input = !check_for_nested_inputs(params);
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:325]: Nit: since we are just using this as a helper now and not a filter_function we could probs switch this logic, but no worries

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 8]
```diff
@@ -246,27 +314,77 @@ inline bool check_tensor_shapes(sdp_params params, bool debug) {
...
+                                                          k_batch_size,
+                                                          v_batch_size,
+                                                          "batch size ",
+                                                          debug);
+
+      // if only one of k or v require broadcasting of batch size, the other
+      // must have a consistent seq_len dim
+      if (broadcastable_batch_size) {
+        if (k_batch_size == 1 && v_batch_size != 1) {
+          const auto v_nt_tensor_impl = at::native::get_nested_tensor_impl(params.value);
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:344]: maybe pull this little chunk out and apply on v and then k if needed

File: test/test_nestedtensor.py [hunk 9]
```diff
@@ -1861,27 +1862,28 @@ def rand_mask(size):
...
+            expected_output_nested = torch.nested.nested_tensor(expected_outputs)
+            self.assertEqual(actual, expected_output_nested)
+
+            # Error case: explicit attn_mask set.
+            with self.assertRaisesRegex(RuntimeError, "not supported when an explicit attn_mask is set"):
+                torch.nn.functional.scaled_dot_product_attention(
+                    query, key, value, attn_mask=attn_mask, dropout_p=dropout_p)
+
+            # Error case: is_causal=True.
+            with self.assertRaisesRegex(RuntimeError, "not supported when is_causal=True"):
```
Comment (drisspg) [inline test/test_nestedtensor.py:1884]: I want to check this, cause I don't know if this fails with the fused kernels

**Review by drisspg — APPROVED** [review]
Awesome!

File: test/test_nestedtensor.py (hunk 9 above)
Comment (mikaylagawarecki) [reply test/test_nestedtensor.py:1884]: I think I shouldn't have indented this code block with a `sdp_kernel(enable_flash=False, enable_math=True, enable_mem_efficient=False)`, both num_head and seq_len are ragged here so it should always dispatch to math, but I should still be checking that it doesn't dispatch to the wrong kernel, my bad
